### PR TITLE
cli: add "location" and "branch" flags to be fully automatable

### PR DESCRIPTION
**Pull request by solomoncyj**

### CODE REVIEW

File: cli.go [hunk 1]
```diff
@@ -19,8 +19,15 @@ func main() {
...
 	var uninstallOpenAsar = flag.Bool("uninstall-openasar", false, "Uninstall OpenAsar from a Discord install")
 	var updateFlag = flag.Bool("update", false, "Update your local Vencord files")
+	var dir = flag.string("dir", false, " Select the location of your Discord client")
+	var client = flag.string("client", "default", "Select the branch of Discord you wish to modify [default|stable|ptb|canary]")
 	flag.Parse()
-
+	if *dir != "" && **client != "" {
+		return "", errors.New("the dir and client switches are mutally exclusive")
+	}
+	if *client && ! *client == "default" || *client == "stable" || *client == "ptb" || *client == "canary" ||) {
```
Comment (kalexmills) [inline cli.go:28]: **suggestion:** consider `*client != "default"`, I think it's a bit clearer than `! *client == "default"`.

It also looks to me like you're missing an open parentheses and you've got an extra `||` operator at the end of this line. I don't think this will compile as-is.

File: cli.go [hunk 2]
```diff
@@ -60,39 +67,52 @@ func main() {
...
-			if _, err := fmt.Scan(&custom); err == nil {
-				if discord := ParseDiscord(custom, ""); discord != nil {
-					return discord
-				}
-			}
-		}
-
-		fmt.Println("That wasn't a valid choice")
-	}
+func PromptDiscord(action string, client *DiscordInstall, dir string) *DiscordInstall {
```
Comment (kalexmills) [inline cli.go:70]: **issue:** It looks like all of the callsites of this `func` will also need to be changed. You should find all of them in your IDE and work through how to pass these particular variables down, in order to make this change work out.

File: cli.go [hunk 3]
```diff
@@ -60,39 +67,52 @@ func main() {
...
+            } else {
+                var custom string
+                fmt.Print("Custom Discord Install: ")
+                if _, err := fmt.Scan(&custom); err == nil {
+                    if discord := ParseDiscord(custom, ""); discord != nil {
+                        return discord
+                    }
+                }
+            }
+        }
```
Comment (kalexmills) [inline cli.go:112]: **comment (non-blocking):** It looks like you changed the indentation formatting in this block, which the maintainer may have opinions about.

**Review by kalexmills — COMMENTED** [review]
Was requested to take a look at this by the author on a Discord. I left some comments accordingly.

As is, I'm not sure this will compile without changes.

File: cli.go [hunk 4]
```diff
@@ -19,8 +19,15 @@ func main() {
 	var installOpenAsar = flag.Bool("install-openasar", false, "Install OpenAsar on a Discord install")
 	var uninstallOpenAsar = flag.Bool("uninstall-openasar", false, "Uninstall OpenAsar from a Discord install")
 	var updateFlag = flag.Bool("update", false, "Update your local Vencord files")
+	var dir = flag.string("dir", false, " Select the location of your Discord client")
+	var client = flag.string("client", "default", "Select the branch of Discord you wish to modify [default|stable|ptb|canary]")
```
Comment (Vendicated) [inline cli.go:23]: have you tested this by building and running it? These method names for example should be uppercase: `flag.String` and Alex's feedback below also suggests that this will not work in the current state. You should use an IDE with support for go such as Goland or VSCode (https://code.visualstudio.com/docs/languages/go)

**Comment by solomoncyj** [comment]
current errors:
```
go build --tags cli
# github.com/Vendicated/VencordInstaller
.\cli.go:29:5: invalid operation: *client && (*client != "default" || *client != "stable" || *client != "ptb" || *client != "canary") (mismatched types string and untyped bool)
.\cli.go:43:21: not enough arguments in call to PromptDiscord
        have (string)
        want (string, *DiscordInstall, string)
.\cli.go:45:21: not enough arguments in call to PromptDiscord
        have (string)
        want (string, *DiscordInstall, string)
.\cli.go:49:28: not enough arguments in call to PromptDiscord
        have (string)
        want (string, *DiscordInstall, string)
.\cli.go:56:28: not enough arguments in call to PromptDiscord
        have (string)
        want (string, *DiscordInstall, string)
        ```

**Comment by solomoncyj** [comment]
ok now it will compile, but has bugs

**Comment by solomoncyj** [comment]
ok. my pr is ready

**Comment by solomoncyj** [comment]
@Vendicated can you pls merge this? Thanks

File: cli.go [hunk 5]
```diff
@@ -60,13 +79,30 @@ func main() {
 	}
 }
 
-func PromptDiscord(action string) *DiscordInstall {
+func PromptDiscord(action string, client *DiscordInstall, dir string, branch string) *DiscordInstall {
```
Comment (Vendicated) [inline cli.go:82]: The client param seems unused and you always pass nil to it. Please remove it if unused

```suggestion
func PromptDiscord(action, dir, branch string) *DiscordInstall {
```

File: cli.go [hunk 6]
```diff
@@ -6,10 +6,19 @@ import (
...
 var discords []any
 
+func isAllowedClient(client string) bool {
+	switch client {
+	case "", "default", "stable", "ptb", "canary":
+		return true
+	default:
+		return false
+	}
+}
```
Comment (Vendicated) [inline cli.go:21]: this function needs some room to breathe
```suggestion
}

```

File: cli.go [hunk 7]
```diff
@@ -32,20 +51,20 @@ func main() {
 
 	var err error
 	if *installFlag {
-		_ = PromptDiscord("patch").patch()
+		PromptDiscord("patch", nil, *dir, *client).patch()
 	} else if *uninstallFlag {
-		_ = PromptDiscord("unpatch").unpatch()
+		PromptDiscord("unpatch", nil, *dir, *client).unpatch()
 	} else if *updateFlag {
-		_ = installLatestBuilds()
```
Comment (Vendicated) [inline cli.go:39]: why did you remove these `_ =`? They are there to explicitly discard the potential error returned by these functions and suppress lint errors

image

File: cli.go [hunk 8]
```diff
@@ -60,13 +79,30 @@ func main() {
...
+func PromptDiscord(action string, client *DiscordInstall, dir string, branch string) *DiscordInstall {
+
+	if branch != "" {
+		for _, discord := range discords {
+			install := discord.(*DiscordInstall)
+			if install.branch == branch {
+				return install
+			}
+		}
+		log.Fatal("Discord" + branch + "is not installed on your pc")
```
Comment (Vendicated) [inline cli.go:91]:
```suggestion
		log.Fatal("Discord " + branch + " is not installed on your pc")
```

File: cli.go [hunk 9]
```diff
@@ -19,8 +28,18 @@ func main() {
 	var installOpenAsar = flag.Bool("install-openasar", false, "Install OpenAsar on a Discord install")
 	var uninstallOpenAsar = flag.Bool("uninstall-openasar", false, "Uninstall OpenAsar from a Discord install")
 	var updateFlag = flag.Bool("update", false, "Update your local Vencord files")
+	var dir = flag.String("dir", "", "Select the location of your Discord client")
+	var client = flag.String("client", "", "Select the branch of Discord you wish to modify [default|stable|ptb|canary]")
 	flag.Parse()
 
+	if *dir != "" && *client != "" {
+		log.Fatal("The 'dir' and 'client' flags are mutually exclusive.")
```
Comment (Vendicated) [inline cli.go:36]: please change these usages of the `log` package to `fmt`. log adds pointless timestamps which distract from the actually relevant information
image

**Comment by Vendicated** [comment]
when using the -dir flag, it still prompts you to pick an install, it just adds the passed dir as an additional option. Wouldn't it be better to skip the picker entirely and directly try to parse the install at the specified direction and if failed print an error and exit?

File: cli.go [hunk 10]
```diff
@@ -60,13 +82,35 @@ func main() {
...
+			if install.branch == branch {
+				return install
+			}
+		}
+		fmt.Println("Discord" + branch + " is not installed on your pc")
+		os.Exit(1)
+	} else if *&dir != "" {
+		if discord := ParseDiscord(*&dir, branch); discord != nil {
+			return discord
+		}
```
Comment (Vendicated) [inline cli.go:99]:
```suggestion
		} else {
			fmt.Println("The specified directory is not a valid Discord install")
			os.Exit(1)
		}
```

Comment (Vendicated) [reply cli.go:99]: after this change, you will now have obsolete dir logic at the bottom so please remove it

**Comment by solomoncyj** [comment]
Review pls

File: cli.go [hunk 11]
```diff
@@ -60,12 +82,31 @@ func main() {
...
+		for _, discord := range discords {
+			install := discord.(*DiscordInstall)
+			if install.branch == branch {
+				return install
+			}
+		}
+		fmt.Println("Discord" + branch + " is not installed on your pc")
+		os.Exit(1)
+	}
+	if *&dir != "" {
```
Comment (Vendicated) [inline cli.go:97]:
```suggestion
	if dir != "" {
```
